source/common/thread_local/thread_local_impl.cc

```diff
@@ -73,6 +73,8 @@ void InstanceImpl::removeSlot(SlotImpl& slot) {
 
   const uint64_t index = slot.index_;
   slots_[index] = nullptr;
+  auto ret = free_slot_indexes_.emplace(index);
+  ASSERT(ret.second, fmt::format("slot index {} already in free slot set!", index));
```

Comment:
With std::list you can do a similar assert using find

Reply:
find is again O(N), and in this case it will go over the full list (as there won't be such a idx if we do things right here)
But if we don't care about the assertion, we can use std::list which is also O(1) allocation/return.

Reply:
Yeah I would just switch to list. I don't think the perf of the assertion matters. Thank you!

Reply:
change to std::list. 
For some reason I thought ASSERT won't be compiled out.  :/